openedx_learning/core/components/api.py

```diff
@@ -98,10 +166,84 @@ def create_component_and_version(
         )
         return (component, component_version)
 
+def get_component(component_pk: int) -> Component:
+    """
+    Get Component by its primary key.
+
+    This is the same as the PublishableEntity's ID primary key.
+    """
+    return Component.with_publishing_relations.get(pk=component_pk)
 
-def get_component_by_pk(component_pk: int) -> Component:
-    return Component.objects.get(pk=component_pk)
+def get_component_by_key(
+    learning_package_id: int,
+    namespace: str,
+    type: str,  # pylint: disable=redefined-builtin
+    local_key: str,
+):
+    """
+    Get Componet by it unique namespace/type/local_key tuple.
+    """
+    return Component.with_publishing_relations \
+                    .get(
+                        learning_package_id=learning_package_id,
+                        namespace=namespace,
+                        type=type,
+                        local_key=local_key,
+                    )
+
+def component_exists_by_key(
+    learning_package_id: int,
+    namespace: str,
+    type: str,  # pylint: disable=redefined-builtin
+    local_key: str
+):
+    """
+    Return True/False for whether a Component exists.
+
+    Note that a Component still exists even if it's been soft-deleted (there's
+    no current Draft version for it), or if it's been unpublished.
+    """
+    try:
+        _component = Component.objects.only('pk').get(
+            learning_package_id=learning_package_id,
+            namespace=namespace,
+            type=type,
+            local_key=local_key,
+        )
+        return True
+    except Component.DoesNotExist:
+        return False
+
+def get_components(
+    learning_package_id: int,
+    draft: bool | None = None,
+    published: bool | None = None,
+    namespace: str | None = None,
+    types: list[str] | None = None,
+    title: str | None = None,
+) -> QuerySet:
+    """
+    Fetch a QuerySet of Components for a LearningPackage using various filters.
+
+    This method will pre-load all the relations that we need in order to get
+    info from the Component's draft and published versions, since we'll be
+    referencing these a lot.
+    """
+    qset = Component.with_publishing_relations \
+                    .filter(learning_package_id=learning_package_id) \
+                    .order_by('pk')
+    if draft is not None:
+        qset = qset.filter(publishable_entity__draft__version__isnull = not draft)
+    if published is not None:
+        qset = qset.filter(publishable_entity__published__version__isnull = not published)
+    if namespace is not None:
+        qset = qset.filter(namespace=namespace)
+    if types is not None:
+        qset = qset.filter(type__in=types)
+    if title is not None:
+        qset = qset.filter(publishable_entity__draft__version__title__icontains=title)
```

Comment:
Good point. I'll make `draft_title` and `published_title` as separate fields that can be passed in.

Reply:
Implementing this now. FWIW, I want to avoid any inference here (e.g. `published=True` implying that a title check should happen against the published version only). I figure any smarts like that should be in the layer above, and these query fields will just map one-to-one to know fields–hence separating `title` into separate parameters for draft vs. published titles.